lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,197 @@
+      end
+
+      def self.create_event(name, manual_reset = false, initial_state = false)
+        handle = FFI::Pointer::NULL_HANDLE
+
+        FFI::Pointer.from_string_to_wide_string(name) do |name_ptr|
+          handle = CreateEventW(FFI::Pointer::NULL,
+            manual_reset ? 1 : FFI::WIN32_FALSE,
+            initial_state ? 1 : FFI::WIN32_FALSE,
+            name_ptr)
+
+          if handle == FFI::Pointer::NULL_HANDLE
+            msg = "Failed to create new event #{name}"
+            raise Puppet::Util::Windows::Error.new(msg)
+          end
+        end
+
+        handle
+      end
+
+      WAIT_ABANDONED = 0x00000080
+      WAIT_OBJECT_0 = 0x00000000
+      WAIT_TIMEOUT = 0x00000102
+      WAIT_FAILED = 0xFFFFFFFF
+
+      def self.wait_on(wait_object, timeout_ms = 50)
+        wait_result = Puppet::Util::Windows::Process::WaitForSingleObject(
+          wait_object, timeout_ms)
+
+        case wait_result
+        when WAIT_OBJECT_0
+          Puppet.debug "Wait object signaled"
+        when WAIT_TIMEOUT
+          Puppet.debug "Waited #{timeout_ms} milliseconds..."
+        # only applicable to mutexes - should never happen here
+        when WAIT_ABANDONED
+          msg = 'Catastrophic failure: wait object in inconsistent state'
+          raise Puppet::Util::Windows::Error.new(msg)
+        when WAIT_FAILED
+          msg = 'Catastrophic failure: waiting on object to be signaled'
+          raise Puppet::Util::Windows::Error.new(msg)
+        end
+
+        wait_result
+      end
+
+      def write_stdin(input)
+        @stdin.puts(input)
+      rescue => e
+        msg = "Error writing STDIN / reading STDOUT: #{e}"
+        raise Puppet::Util::Windows::Error.new(msg)
+      end
+
+      def drain_stdout
+        output = []
+        while self.class.is_readable?(@stdout, 0.1) do
+          l = @stdout.gets
+          Puppet.debug "#{Time.now} STDOUT> #{l}"
+          output << l
+        end
+        output
+      end
+
+      def read_stdout(output_ready_event, timeout_ms = 20 * 1000, wait_interval_ms = 50)
+        output = []
+        waited = 0
+
+        # drain the pipe while waiting for the event signal
+        while WAIT_TIMEOUT == self.class.wait_on(output_ready_event, wait_interval_ms)
+          output << drain_stdout
+          waited += wait_interval_ms
+          if (waited > timeout_ms)
+            msg = "Catastrophic failure: wait object timeout #{timeout_ms} exceeded"
+            raise Puppet::Util::Windows::Error.new(msg, WAIT_TIMEOUT)
+          end
+        end
+
+        Puppet.debug "Waited #{waited} total milliseconds."
+
+        # once signaled, ensure everything has been drained
+        output << drain_stdout
+
+        output.join('')
+      rescue => e
+        msg = "Error reading STDOUT: #{e}"
+        raise Puppet::Util::Windows::Error.new(msg)
+      end
+
+      def exec_read_result(powershell_code, output_ready_event, timeout_ms = 120 * 1000)
```

Comment:
Is there anything that would take more than 5 minutes? I am thinking more like 10 or longer. Choco does a default timeout on commands at a ludicrous default of 45 minutes. That way if there is a download on a slow connection or if the install takes forever, it has a chance to finish.

I'm thinking something like 5 -10 minutes?